### PR TITLE
LPS-120270 Refactor AuthVerifierPipeline and use URLPatternMapper

### PR DESCRIPTION
**Pull request by csierra**
Hi @shuyangzhou, /cc @arthurchan35

we have give it another twist after your suggestions in our last pull. We were holding back some other refactors but we decided to byte the bullet and make them now since we believe as well the result is better. 

In this pull we should have fixed the _"double tracking issues"_ that we had in the previous one, while we keep all the backwards compatibility with previous behavior. There are some semantics that should also become clerare now, such as:

* *AuthVerifier instances should be singletons*: With current `AuthVerifier` design they are expected to be singletons. However they were instantiated several times in order to register new configurations to the portal pipeline. This behavior was ambiguous and prevented `AuthVerifier` implementations to be overriden. We have turned `AuthVerifier` instances into singletons and allow for overriding. Also we allow registering `AuthVerifierConfiguration` instances to the portal pipeline. 
* There is a clear distinction between portal and modules pipelines: and modules also apply portal pipeline, since it is the one configurable for portal settings. 
* We have added an `AuthVerifierRegistry` that maintains all the booking and allows for `AuthVerifier` _lazy_ fetching so we don't need to keep a double tracking.

Please let us know what you think... but don't forward to Brian yet since we might perform additional cleaning. 

Thanks un advance. 

Carlos.

### CODE REVIEW

File: portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierFilter.java [hunk 1]
```diff
@@ -202,6 +220,61 @@ else if (_guestAllowed || (state == AuthVerifierResult.State.SUCCESS)) {
 		}
 	}
 
+	private List<AuthVerifierConfiguration> _buildAuthVerifierConfigurations(
+		Map<String, Object> initParametersMap) {
+
+		Map<String, Integer> authVerifierConfigurationIndexs = new HashMap<>();
```
Comment (shuyangzhou) [inline portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierFilter.java:226]: @csierra this map has never been added with anything.

Comment (csierra) [reply portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierFilter.java:226]: you are right... I actually get a warning from the IDE but didn't spot it... @arthurchan35 would you mind having a look?

**Comment by shuyangzhou** [comment]
@csierra this pull is still too big and complex for me to do any detailed level review. If that is needed, the pull needs to be broke down into smaller piece with more tests to show what is expected. Otherwise I would only be able to stay at the superficial level to catch obvious mistakes, that won't be much helpful.

**Comment by csierra** [comment]
@shuyangzhou, /cc @arthurchan35 

we are sending this to you because Brian asked us to send it to you back in the days, and also we appreciate all the eyes we can borrow. 
in the previous pull you already commented that the double tracking we were making wasn't great (and we concurred) so it is being helpful. 

Might it help if we explain in more detail what we are after and how we are trying to achieve it, wither with docs or tetsts? Maybe we can come up with more tests, thought this is mainly a refactor keeping the old behavior, so current tests must work (except the unit test one that's using mocks). 

Would it help if we shared a screen on your office time? it should be possible for wither Arthur or me to do that on a time that is comfortable to you. 

Let's us know your thoughts. 

Carlos.

File: portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierFilter.java (hunk 1 above)
Comment (arthurchan35) [reply portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierFilter.java:226]: Looking

**Comment by shuyangzhou** [comment]
@csierra the best way for me to digest this is by smaller pieces or more tests. I don't have enough background knowledge on the business logic, it is better for me to stay that way to only focus on what is expected and what it is actually happening. If that is too much work or not really worthy for you guys, I can still take it in as a whole, just it will be harder for me.

**Comment by arthurchan35** [comment]
@shuyangzhou ,
I will try to split up the commits into smaller pieces, and add more tests tomorrow.
And we try to add pull comments to explain the structure and reason behind it.

**Comment by arthurchan35** [comment]
ci:close

**Comment by shuyangzhou** [comment]
Thanks @arthurchan35 !

**Pull request closed without merging**